include/unifex/io_concepts.hpp

```diff
@@ -0,0 +1,201 @@
+ * you may not use this socket except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/tag_invoke.hpp>
+
+// The senders would be loosened eventually to support sequences.
+//
+// // make async_read_some act like blocking read()
+// auto bytes_read = sync_wait(
+//     select_first_arg(
+//         async_read_some(socket)(just(buffer)))).size();
+//
+// // make async_write_some act like blocking write()
+// auto bytes_written = sync_wait(
+//     select_first_arg(
+//         async_write_some(socket)(just(buffer)))).size();
+//
+// // echo
+// sync_wait(
+//     repeat(async_write_some(socket)(
+//         select_first_arg(
+//             async_read_some(socket)(just(buffer))))));
+//
+// // echo with pipe
+// just(buffer) |
+//     async_read_some(socket) |
+//     select_first_arg() |
+//     async_write_some(socket) |
+//     repeat() |
+//     sync_wait();
+//
+
+namespace unifex {
+namespace _io_cpo {
+//
+// async_read_some
+//
+// returns a function that takes a sender of BufferSequence
+// which returns a sender of 2 BufferSequence. One that has been truncated
+// to only describe the filled portions of the original BufferSequence
+// Another that only describes the remaining portions of the original
+// BufferSequence
+//
+inline constexpr struct async_read_some_cpo {
+  template <typename ForwardReader, typename BufferSequence>
+  auto operator()(
+      ForwardReader& socket,
+      BufferSequence&& bufferSequence) const
+      noexcept(is_nothrow_tag_invocable_v<
+               async_read_some_cpo,
+               ForwardReader&,
+               BufferSequence>)
+          -> tag_invoke_result_t<
+              async_read_some_cpo,
+              ForwardReader&,
+              BufferSequence> {
+    return unifex::tag_invoke(
+        *this, socket, (BufferSequence &&) bufferSequence);
+  }
+
+  template <typename ForwardReader>
+  auto operator()(
+      ForwardReader& socket) const
+      noexcept(is_nothrow_tag_invocable_v<
+               async_read_some_cpo,
+               ForwardReader&>)
+          -> tag_invoke_result_t<
+              async_read_some_cpo,
+              ForwardReader&> {
+    return unifex::tag_invoke(*this, socket);
+  }
+} async_read_some{};
+
+// async_write_some
+//
+// returns a function that takes a sender of BufferSequence
+// which returns a sender of 2 BufferSequence. One that has been truncated
+// to only describe the written portions of the original BufferSequence
+// Another that only describes the remaining portions of the original
+// BufferSequence
```

Comment:
I do not believe that an already available buffer describes the basis io operation. The buffer should be supplied by a predecessor.

Reply:
You may wish to consider, when considering how to support ASIO-like BufferSequences, https://github.com/ned14/llfio/blob/resumable_io_branch_merge/include/llfio/v2.0/io_handle.hpp#L180.

(In short: don't do what ASIO does, it's highly suboptimal)

Reply:
@ned14, do you have a design in mind?

Reply:
@kirkshoop When I pitched to WG21 the proposal that each handle type declares a `buffers_type` native to it, that was very warmly received. There was a strong consensus in both the room and outside the room that this was the correct approach, as on some platforms the native scatter-gather buffer type differs between handle implementations. ASIO's buffer sequence adapter framework would sit separately, and above this, as a higher level abstraction, and for backwards compatibility with what ASIO currently does. So don't write code in terms of BufferSequence, wrong abstraction. Write code in terms of handle native buffer types, for some templated handle type.

Additionally, the specific line number I linked you to shows the API `result<registered_buffer_type> allocate_registered_buffer(size_t &bytes)`. It goes and asks your chosen i/o multiplexer (e.g. Windows RIO) for a registered i/o buffer and returns it. The allocation is type erased into a `shared_ptr`, so external code need care nothing about Windows RIO, and when the shared ptr count reaches zero, the buffer is recycled/deregistered.

Bringing things back to pipes, you can make pipes and socket i/o on Linux go extremely quickly indeed with registered buffers. Just `vmsplice()` the pages backing the registered i/o buffer directly into the pipe (and splice from there into any sockets). Zero memory copying. Basically the same as Windows RIO. Note that the LLFIO API design allows both options, whilst user code never notices the implementation detail.

That's the correct design, and that's what I'd recommend to you.

Reply:
I like the ability to allocate the buffer. I think that buffer allocation is non-blocking/async. If I have a tuned pool of buffers I am allocating from, or splicing into, then I do not want to block threads waiting for a buffer to free.

Reply:
Registered buffer allocation *may* be fast, or it may be very slow. Depends on the multiplexer implementation - certainly for Windows RIO, it locks buffers into memory, and that is subject to the 128Kb per-process memory locking limit. So the ability to build a tuned pool of buffers is very limited, plus is a process-wide limit.

For some use cases, whole 4Kb or 2Mb page granularity i/o buffers is acceptable. For others (specifically when you must pack data of unknown lengths during writes), you work in terms of long gather write buffer sequences. You need to support **both** use cases.

Just to support what I just said, I'm not sure if Eric told you what I do for my day job. I'm helping build out a database which captures all trades which occur in the US, live, and lets your run queries upon them. Data ingress is about ~Gb per second per node. The whole thing runs on LLFIO.

We write that data by a very long gather buffer list consisting of many variably sized buffers all over memory. This is the least worst choice available to us, as the incoming data must be indexed during ingress, and it is unavoidably broken up over memory. For reads, however, we do all i/o in whole 4Kb or 2Mb page sized amounts, and we manually kick those pages in and out of memory, always passing around views onto that memory through C++.

Anyway, my point is that there is no one sized fits solution for all. I would not make decisions on behalf of users about buffer management, for sure. I would offer whatever the OS kernel offers, and let users choose their poison. Even if that blocks threads waiting for a buffer to free.

Reply:
The whole point of sender/receiver is to allow more implementations not limit them. 

when allocation is expressed as a blocking api the only way to respond to limits and scheduling is to block a thread or fail with an error. 

when allocation is expressed as a sender/coroutine then an implementation is allowed to block a thread, fail with an error and suspend until the resource becomes available.

Reply:
@kirkshoop I think we must have crossed wires here. You don't register nor allocate i/o buffers during i/o, that would be suicidal for performance. You *register* i/o buffers before you ever do i/o, usually in `main()` or other cold code. You then are required to only ever use scatter gather buffers from within those pre-registered i/o buffers. You can see the relevant API spec at https://github.com/ned14/llfio/blob/resumable_io_branch_merge/include/llfio/v2.0/io_handle.hpp#L247:

```c++
io_result<buffers_type>
read(registered_buffer_type base,
     io_request<buffers_type> reqs,
     deadline d = deadline()) noexcept;
```
For further information, please see:

- [`RIORegisterBuffer`](https://docs.microsoft.com/en-us/windows/win32/api/mswsock/nc-mswsock-lpfn_rioregisterbuffer)
- [`RIOReceive`](https://docs.microsoft.com/en-us/windows/win32/api/mswsock/nc-mswsock-lpfn_rioreceiveex)

RIO is io_uring for Windows, similar shared ringbuffer based design except sockets only. If Windows does go ahead and implement a generalised io_uring equivalent for all Windows kernel i/o, I am told it will look very much like RIO.